transport/src/main/java/io/netty/bootstrap/DynamicLocalSocketAddress.java

```diff
@@ -0,0 +1,123 @@
+/*
+ * Copyright 2019 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.bootstrap;
+
+import io.netty.util.internal.ObjectUtil;
+
+import java.net.Inet4Address;
+import java.net.Inet6Address;
+import java.net.InetAddress;
+import java.net.InetSocketAddress;
+import java.net.NetworkInterface;
+import java.net.SocketAddress;
+import java.util.Arrays;
+import java.util.Enumeration;
+import java.util.Iterator;
+
+/**
+ * {@link SocketAddress} which will provide the local {@link SocketAddress} to bind to after the remote address
+ * is resolved and so can be used with {@link Bootstrap#connect(SocketAddress, SocketAddress)}.
+ */
+public abstract class DynamicLocalSocketAddress extends SocketAddress {
+
+    /**
+     * Return a {@link SocketAddress} to bind to for the given {@code remote} {@link SocketAddress} or {@code null}
+     * if no specific should be used.
+     */
+    public abstract SocketAddress address(SocketAddress remote);
```

Comment:
How about moving this method below the static factory methods?

Reply:
I actually removed all the implementations for now.